### ISSUE TITLE
BABEL_PARSER_SYNTAX_ERROR error with 2.3.1 when using d3

### ISSUE THREAD

**Issue by pfafman** [meteor/meteor#11525]
Getting BABEL_PARSER_SYNTAX_ERROR when starting up meteor on 2.2.1, 2.3 or 2.3.1.  Report here is for 2.3.1

Steps to reproduce:

1. create a todo app with blaze:
2.

**Comment by StorytellerCZ** [comment]
Is it the same as before or is it coming from a new package? Sadly the output here isn't of much help.

**Comment by StorytellerCZ** [comment]
Probably again this: https://github.com/d3/d3-array/issues/216

**Comment by StorytellerCZ** [comment]
This issue tracks it in Babel: https://github.com/babel/babel/issues/11038

**Comment by pfafman** [comment]
I saw the babel issue but looks like no one has done anything in 6 months.  The thing I am trying to tack down is why any version prior 20 meteor 2.2.1 works

As stated this is just the example code with d3 added in.  No new package and the output is all babel gives me.

**Comment by pfafman** [comment]
Any of my projects with d3 gets the same error.  If someone is using d3 with meteor 2.2.1 or greater and it is working would like to know.

**Comment by StorytellerCZ** [comment]
:thinking: I think that it could be either babel/typescript updates or maybe one of the babel related dependencies. We should probably try to play with those for now.

**Comment by pfafman** [comment]
Downgrading to d3 v5 removes the error when using meteor 2.2.1 -2.3.1

Downgrading meteor to 2.2 removes the error when using d3 vs 7

Commenting out the lines in the `cumsum.js` file in d3 which is where the error is coming from.
or editing it to make it a little more readable:

```
export default function cumsum(values, valueof) {
  var sum = 0, index = 0;
  return Float64Array.from(values, (v) => {
    if (valueof === undefined) {
      (sum += +v || 0)
    } else {
      (sum += +valueof(v, index++, values) || 0)
    }
  });
}
```
 
Since the babel npm version is the same in all the above combinations I think it is how the newer versions of meteor are handing the code to babel???

**Comment by bluefangs** [comment]
I hit upon a very similar error with @amcharts/maps. Downgrading to 2.2 seems to solve the issue for now.

**Comment by derwaldgeist** [comment]
Unfortunately, that affects any library that is using D3 under the good. We’re using react-globe-gl and fell into the same trap when upgrading Meteor to 2.2.1.

**Comment by StorytellerCZ** [comment]
Looks like they have made another go at fixing this: https://github.com/babel/babel/pull/13581

**Comment by StorytellerCZ** [comment]
I'm guessing it will be in Babel 7.14.9. I will update all the packages when it comes out.

**Comment by derwaldgeist** [comment]
Thanks. Any chance to get this down ported to 2.2.1? I’m hesitating to upgrade to Node 14.

**Comment by StorytellerCZ** [comment]
@derwaldgeist I'll look into it once the fix comes out.

**Comment by StorytellerCZ** [comment]
I have updated babel for 2.3.3 both in the bundle and in skeletons. I'm going to the new version now to see if it resolves the issue.

**Comment by larswolter** [comment]
Hi, at least for me it does not solve the issue. It seems that the meteor [babel-compiler package ](https://github.com/meteor/meteor/blob/6f7df8f35f1617b393a1d13c48f2bf70c2be31f6/packages/babel-compiler/package.js#L13) uses [@meteorjs/babel wrapper](https://github.com/meteor/babel/blob/16c29604beb40a0f6c4a79f98d1bea70269711d0/package.json#L34) with @babel/parser@7.14.3
I think this needs to be updated, or is there a way to tell the compiler to use the babel npms installed in the project?

**Comment by StorytellerCZ** [comment]
@larswolter Good catch! I will update that one asap!

**Comment by StorytellerCZ** [comment]
Even with all of that I'm still getting the error in `2.4-beta.8`. Did I miss something?

**Comment by pfafman** [comment]
I upgrade my test app of this (modified simple-tofod-blaze with d3  added) and I am still getting the error  about the cumsum.js file  but it is a little nicer error output. This was even after removing and reinstalling all the NPMs.

```
simple-todos-blaze$ meteor


[[[[[ ~/Projects/coWorkr/simple-todos-blaze ]]]]]

=> Started proxy.
=> Started HMR server.
=> Errors prevented startup:

   While building for web.browser:
   node_modules/d3-array/src/cumsum.js:4:15: Unexpected token (4:15)

   While building for web.browser.legacy:
   node_modules/d3-array/src/cumsum.js:4:15: Unexpected token (4:15)

=> Your application has errors. Waiting for file change.
=> Started MongoDB.
```

**Comment by larswolter** [comment]
For me it also fails with the new ersion but without the nice error output. I tried to verify if the new @babel/parser fixes the problem and it does not. The error only occurs when using the babel/parser flow plugin, following is a small example to test it
```javascript
import babelParser from '@babel/parser';
import fs from 'fs';

// meteor default (from Meteor babel-compiler Babel.getDefaultOptions()))
//const options = { "allowImportExportEverywhere": true, "allowReturnOutsideFunction": true, "allowUndeclaredExports": true, "plugins": ["*", "flow", "jsx", "asyncGenerators", "bigInt", "classPrivateMethods", "classPrivateProperties", "classProperties", "doExpressions", "dynamicImport", "exportDefaultFrom", "exportExtensions", "exportNamespaceFrom", "functionBind", "functionSent", "importMeta", "nullishCoalescingOperator", "numericSeparator", "objectRestSpread", "optionalCatchBinding", "optionalChaining", ["pipelineOperator", { "proposal": "minimal" }], "throwExpressions"], "sourceType": "module", "strictMode": false }

// meteor default (from Meteor babel-compiler Babel.getDefaultOptions())) without flow
//const options = { "allowImportExportEverywhere": true, "allowReturnOutsideFunction": true, "allowUndeclaredExports": true, "plugins": ["*", "jsx", "asyncGenerators", "bigInt", "classPrivateMethods", "classPrivateProperties", "classProperties", "doExpressions", "dynamicImport", "exportDefaultFrom", "exportExtensions", "exportNamespaceFrom", "functionBind", "functionSent", "importMeta", "nullishCoalescingOperator", "numericSeparator", "objectRestSpread", "optionalCatchBinding", "optionalChaining", ["pipelineOperator", { "proposal": "minimal" }], "throwExpressions"], "sourceType": "module", "strictMode": false }

// minimum failing
const options = { sourceType: 'module', plugins: ['flow'] };
const res = babelParser.parse(
  fs.readFileSync('./node_modules/d3-array/src/cumsum.js').toString(),
  options);
console.log(`Successfully parsed src`);
```
Using the following package.json
```json
{
  "name": "parsercheck",
  "version": "1.0.0",
  "description": "test for d3-array compiling failure",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Lars Wolter",
  "license": "ISC",
  "dependencies": {
    "@babel/parser": "^7.14.9",
    "d3-array": "^3.0.1"
  }
}
```

**Comment by larswolter** [comment]
The according issue in Babel https://github.com/babel/babel/issues/13644

**Comment by nicolo-ribaudo** [comment]
PR to fix this: https://github.com/babel/babel/pull/13655

**Comment by StorytellerCZ** [comment]
@nicolo-ribaudo thanks for finding the new PR. I'll be watching it and update dependencies accordingly once it is merged.

**Comment by Gywem** [comment]
Grettings @StorytellerCZ!

Great that the fix is coming! 😄 We have faced this issue on a large app that uses d3 as well. Anyway, is there a way to force `babel/parser` version on an existing meteor version for development purposes? Or do we need to wait for the next release?

**Comment by nicolo-ribaudo** [comment]
If you are using Yarn, you can add this to your `package.json`:
```json
"resolutions": {
  "@babel/parser": "1.2.3"
}
```
where `1.2.3` is a known working version.

**Comment by Gywem** [comment]
@nicolo-ribaudo Thank you! I used this on the past and worked well.

However, I think the issue mentioned in this thread is related with meteor infrastructure to bundle and the usage of `@babel/parser` is on the framework level, and if I am not mistaken it is used on several Meteor transpiler libaries and wrappers on the core, like reify, meteor-babel, and maybe others I am not aware of.

So my question is more around how we can relace on-fly the definitions of these libraries on Meteor level if possible. I don't think we have a way to use NPM resolutions straight on Meteor core since NPM packages are described behind the definitions of athmosphere packages and these don't support resolutions. I can only think on use a custom Meteor, and fork all libraries using `@babel/parser` to tweak properly the working version when available. Asking just in case there is a known way on the Meteor development that I am missing.

Anyway, I will wait for the next patch version when ready. 😄

**Comment by StorytellerCZ** [comment]
It is as @Gywem says. I'm watching babel daily to for new versions and will update them for each beta until we get the fix.

**Comment by nicolo-ribaudo** [comment]
Released! https://github.com/babel/babel/releases/tag/v7.15.3

**Comment by StorytellerCZ** [comment]
@nicolo-ribaudo I will make a new 2.4 beta with the new version straight away.

**Comment by Gywem** [comment]
@StorytellerCZ: Can we get a patch on 2.3.x? Or should we upgrade straight to 2.4?

**Comment by StorytellerCZ** [comment]
@Gywem we'll test on 2.4 beta and then I'll look into backporting it to 2.3. There will be a new version anyhow today or tomorrow due to Node security release.

**Comment by derwaldgeist** [comment]
Any chance to downport to 2.2.1 as well?  We're still on this version, since 2.3 brought a new major Node release.

**Comment by StorytellerCZ** [comment]
@derwaldgeist since Node 12.x is also getting a security release I will release 2.2.3 which will include these fixes as well, but 2.3.5 will have priority.

**Comment by derwaldgeist** [comment]
Awesome, thanks!

**Comment by larswolter** [comment]
2.3.5-beta.0 solved it for me. Thanks for the quick fix and backporting!
